### ISSUE TITLE
Missing Batch Info during Returnn Network construction via NNetConstructor 

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#193]

**Comment by albertz** [comment]
A pure RETURNN test case, which is somewhat artificial maybe, and not sure if this is supposed to work:
```python
def test_engine_get_network_data_ref_batch_info():
  from returnn.tf.util.data import Data, BatchInfo, SpatialDim, FeatureDim, batch_dim
  from returnn.config import Config
  from returnn.tf.engine import Engine
  from returnn.datasets import init_dataset
  time_dim = SpatialDim("time")
  in_dim = FeatureDim("in", 3)
  x = Data("data", dim_tags=[batch_dim, time_dim, in_dim], available_for_inference=True)

  def _config_get_network(epoch: int, **_kwargs) -> dict:
    epoch  # unused  # noqa
    if not x.batch:
      x.batch = BatchInfo.make_global_batch_info(-1)
    x.copy()  # this causes the exception
    return {
      "output": {"class": "copy", "from": "data"}
    }

  config = Config({
    "task": "train", "num_epochs": 1, "start_epoch": 1,
    "get_network": _config_get_network,
    "extern_data": {x.name: {"dim_tags": [batch_dim, time_dim, in_dim], "available_for_inference": True}},
  })
  train_dataset = init_dataset(
    {"class": "DummyDataset", "input_dim": in_dim.dimension, "output_dim": 5, "num_seqs": 10})
  engine = Engine(config)
  engine.init_train_from_config(config, train_data=train_dataset)
  engine.train()
```
Traceback:
```
test_TFEngine.py:827: in test_engine_get_network_data_ref_batch_info
    engine.train()
../returnn/tf/engine.py:1449: in train
    self.init_train_epoch()
../returnn/tf/engine.py:1469: in init_train_epoch
    new_network_desc = self.get_net_dict_for_epoch(epoch=self.epoch)
../returnn/tf/engine.py:1071: in get_net_dict_for_epoch
    net_dict = self.custom_get_net_dict(epoch=epoch)
test_TFEngine.py:813: in _config_get_network
    x.copy()
../returnn/tf/util/data.py:3166: in copy
    data = Data(**self.get_kwargs())
../returnn/tf/util/data.py:2778: in __init__
    self.sanity_check(assume_complete=False)
../returnn/tf/util/data.py:2876: in sanity_check
    assert tag.batch == tag.dyn_size_ext.batch
E   AssertionError
      tag = <local> Dim{'time'[B]}
      tag.batch = <local> BatchInfo{B(-1)}
      tag.dyn_size_ext = <local> Data{'data_dim0_size', [B], dtype='int32'}
      tag.dyn_size_ext.batch = <local> BatchInfo{B}
```

**Comment by albertz** [comment]
From this simpler pure RETURNN test case, I think the problem is clear.

We create some dummy batch info `BatchInfo.make_global_batch_info(-1)`.
`x.batch = ...` will also set `time_dim.batch`.

The `ExternData.init_batch_info` logic (here inlined) sets the new batch info on all data (from `extern_data`) and all their dim tags. I think this is fine and makes sense. It does not touch `x` here (also fine).

After that, `time_dim.batch` is again changed to the new batch info.

This is a problem, because `x.dim_tags` still refers to that same `time_dim` object but `x.batch` refers to the initial dummy batch info. So the batch info is not consistent anymore.

A further call like `x._adapt_batch_consistent_dim_tags()` might fix this, but this is not really an option, we want to have the user API also such that the user should not need to care about this.

I'm not really sure what part to blame here. The `ExternData.init_batch_info` logic? The logic of `Data.batch` assignments? The logic of batch consistent dim tag handling in general? Related is https://github.com/rwth-i6/returnn/issues/975.
